src/OpenLoco/src/GameCommands/CreateIndustry.cpp

```diff
@@ -0,0 +1,767 @@
+        std::vector<Colour> availableColours;
+        auto colourBitSet = indObj->availableColours;
+        for (auto colourI32 = Numerics::bitScanForward(colourBitSet); colourI32 != -1; colourI32 = Numerics::bitScanForward(colourBitSet))
+        {
+            colourBitSet &= ~(1ULL << colourI32);
+            availableColours.push_back(static_cast<Colour>(colourI32));
+        }
+
+        // 0x00E0C3D3
+        const auto randColour = [&availableColours, &newIndustry]() {
+            if (availableColours.empty())
+            {
+                return Colour::black;
+            }
+            return availableColours[newIndustry->prng.randNext(availableColours.size() - 1)];
+        }();
+
+        // 0x00E0C3BE - C0
+        auto lastPlacedBuildingPos = World::Pos2{ newIndustry->x, newIndustry->y };
+
+        // used also for 0x00454552 break up into two when function allowed to diverge
+        const auto randVal = newIndustry->prng.randNext() & 0xFF;
+
+        auto prodRateRand = randVal;
+        for (auto i = 0; i < 2; ++i)
+        {
+            newIndustry->var_17D[i] = 0;
+            newIndustry->productionRate[i] = (((indObj->initialProductionRate[i].max - indObj->initialProductionRate[i].min) * prodRateRand) / 256) + indObj->initialProductionRate[i].min;
+            if (isEditorMode())
+            {
+                newIndustry->var_17D[i] = newIndustry->productionRate[i];
+                newIndustry->producedCargoQuantityPreviousMonth[i] = newIndustry->var_17D[i] * 30;
+            }
+            // This is odd but follows vanilla
+            prodRateRand = newIndustry->productionRate[i] & 0xFF;
+        }
+
+        currency32_t totalCost = 0;
+
+        // 0x00454552
+        const auto numBuildings = (((indObj->maxNumBuildings - indObj->minNumBuildings + 1) * randVal) / 256) + indObj->minNumBuildings;
+        for (auto i = 0U; i < numBuildings; ++i)
+        {
+            const auto building = indObj->buildings[i];
+            // 0x00E0C3D2 (bit 0)
+            const bool isMultiTile = indObj->buildingSizeFlags & (1ULL << building);
+
+            bool hasBuildingPlaced = false;
+            // Attempt to place
+            for (auto j = 0U; j < 25; ++j)
+            {
+                const auto randVal2 = newIndustry->prng.randNext();
+                // bh
+                const uint8_t direction = indObj->hasFlags(IndustryObjectFlags::notRotatable) ? 0 : randVal2 & 0x3;
+
+                const auto [distanceRange, minDistance] = [j]() -> std::pair<tile_coord_t, tile_coord_t> {
+                    if (j == 0)
+                    {
+                        return { 1, 0 };
+                    }
+                    else if (j < 8)
+                    {
+                        return { 3, -1 };
+                    }
+                    else
+                    {
+                        return { 5, -2 };
+                    }
+                }();
```

Comment:
This might be also better as a standalone static function, with j being the argument rather than captured.